### PR TITLE
added msgValue feature

### PR DESCRIPTION
**Pull request by tHeMaskedMan981**
- added msgValue support 
- added execution fees calculation based on msgValue  
- updated SocketSrc, SocketDst and Hasher to use messageDetails to reduce variables
- passing payloadSize in getMinFees of ExecutionManager for future fees calculation. 
- added extraParams as a bytes32 parameter which can be used later on for extra instructions.

### CODE REVIEW

File: contracts/ExecutionManager.sol [hunk 1]
```diff
@@ -35,7 +49,21 @@ contract ExecutionManager is IExecutionManager, AccessControlExtended {
...
+        public relativeNativeTokenPrice;
+
+    // mapping(uint32 => uint256) public baseGasUsed;
+
+    mapping(uint32 => mapping(uint32 => uint256)) public msgValueMinThreshold;
+
+    mapping(uint32 => mapping(uint32 => uint256)) public msgValueMaxThreshold;
+
+    // msg.value*scrNativePrice >= relativeNativeTokenPrice[srcSlug][destinationSlug] * destMsgValue /10^18
+
```
Comment (arthcp) [inline contracts/ExecutionManager.sol:63]: srcSlug will always be same for all these right?

File: contracts/ExecutionManager.sol [hunk 2]
```diff
@@ -90,10 +118,23 @@ contract ExecutionManager is IExecutionManager, AccessControlExtended {
...
-        uint256,
+        uint256 gasLimit_,
+        uint256 msgValue_,
+        uint256 payloadSize_,
+        bytes32 extraParams_,
         uint32 siblingChainSlug_
     ) external view override returns (uint256) {
-        return executionFees[siblingChainSlug_];
+        if (msgValue_ == 0) return executionFees[siblingChainSlug_];
+
```
Comment (arthcp) [inline contracts/ExecutionManager.sol:128]: min value check is there so zero check not needed

File: contracts/ExecutionManager.sol [hunk 3]
```diff
@@ -90,10 +118,23 @@ contract ExecutionManager is IExecutionManager, AccessControlExtended {
...
+        if (msgValue_ < msgValueMinThreshold[chainSlug][siblingChainSlug_])
+            revert MsgValueTooLow();
+        if (msgValue_ > msgValueMaxThreshold[chainSlug][siblingChainSlug_])
+            revert MsgValueTooHigh();
+
+        uint256 msgValueRequiredOnSrcChain = (relativeNativeTokenPrice[
+            chainSlug
+        ][siblingChainSlug_] * msgValue_) / (10 ** 18);
```
Comment (arthcp) [inline contracts/ExecutionManager.sol:136]: verify if 10 ** 18 is calculated at run time. if yes lets change to 1e18.

File: contracts/ExecutionManager.sol [hunk 4]
```diff
@@ -90,10 +118,23 @@ contract ExecutionManager is IExecutionManager, AccessControlExtended {
      * @return Minimum fees required for executing the transaction
      */
     function getMinFees(
-        uint256,
+        uint256 gasLimit_,
+        uint256 msgValue_,
+        uint256 payloadSize_,
+        bytes32 extraParams_,
```
Comment (arthcp) [inline contracts/ExecutionManager.sol:124]: extraParams_ is not used anywhere from what I see.
The intent was to decode msgValue_ from extraParams_ and not have it as a dedicated input.

File: contracts/OpenExecutionManager.sol [hunk 5]
```diff
@@ -3,55 +3,33 @@ pragma solidity 0.8.7;
...
 import "./libraries/RescueFundsLib.sol";
 import "./libraries/SignatureVerifierLib.sol";
 import "./libraries/FeesHelper.sol";
-import "./libraries/SignatureVerifierLib.sol";
+import "./utils/AccessControlExtended.sol";
 import {WITHDRAW_ROLE, RESCUE_ROLE, GOVERNANCE_ROLE, EXECUTOR_ROLE, FEES_UPDATER_ROLE} from "./utils/AccessRoles.sol";
-import {FEES_UPDATE_SIG_IDENTIFIER} from "./utils/SigIdentifiers.sol";
+import {FEES_UPDATE_SIG_IDENTIFIER, RELATIVE_NATIVE_TOKEN_PRICE_UPDATE_SIG_IDENTIFIER, MSG_VALUE_MAX_THRESHOLD_SIG_IDENTIFIER, MSG_VALUE_MIN_THRESHOLD_SIG_IDENTIFIER} from "./utils/SigIdentifiers.sol";
+import "./ExecutionManager.sol";
+
```
Comment (arthcp) [inline contracts/OpenExecutionManager.sol:14]: We can remove all extra imports from here

File: contracts/OpenExecutionManager.sol [hunk 6]
```diff
@@ -3,55 +3,33 @@ pragma solidity 0.8.7;
...
-    ) AccessControlExtended(owner_) {
-        signatureVerifier__ = signatureVerifier_;
-        chainSlug = chainSlug_;
-    }
+    ) ExecutionManager(owner_, chainSlug_, signatureVerifier_)
+    {}
 
     /**
-     * @notice this function is open for execution
+     * @notice Checks whether the provided signer address is an executor for the given packed message and signature
```
Comment (arthcp) [inline contracts/OpenExecutionManager.sol:32]: Wrong comment change for OpenExecutionManager

File: contracts/socket/SocketBatcher.sol [hunk 7]
```diff
@@ -209,7 +209,7 @@ contract SocketBatcher is AccessControl {
     function executeBatch(
         address socketAddress_,
         ExecuteRequest[] calldata executeRequests_
-    ) external {
+    ) external payable {
```
Comment (arthcp) [inline contracts/socket/SocketBatcher.sol:212]: value also needs to be passed for each execute call in batch

File: contracts/socket/SocketDst.sol [hunk 8]
```diff
@@ -200,34 +192,32 @@ abstract contract SocketDst is SocketBase {
...
-        uint256 executionFee,
+        address executor_,
         address localPlug_,
         uint32 remoteChainSlug_,
-        uint256 msgGasLimit_,
-        bytes32 msgId_,
-        bytes calldata payload_
+        ISocket.MessageDetails memory messageDetails_
     ) internal {
+        if (msg.value < messageDetails_.msgValue) revert InsufficientMsgValue();
```
Comment (arthcp) [inline contracts/socket/SocketDst.sol:200]: When we change to decoding msgValue from the extraParam, socket will not do this check. The check needs to be done on execution manager where decoding logic is present.

**Review by arthcp — CHANGES_REQUESTED** [review]

File: contracts/ExecutionManager.sol (hunk 2 above)
Comment (tHeMaskedMan981) [reply contracts/ExecutionManager.sol:128]: I kept the 0 check for the case if there is no msgValue, we will just return the execution Fees without any calculation. But if there is some value, we will check min and max values.

Comment (arthcp) [reply contracts/ExecutionManager.sol]: cool makes sense

**Review by arthcp — APPROVED** [review]

**Review by ameeshaagrawal — APPROVED** [review]

**Pull request merged**